core/src/word_lock.rs

```diff
@@ -48,39 +48,39 @@ impl ThreadData {
 }
 
 // Invokes the given closure with a reference to the current thread `ThreadData`.
-fn with_thread_data<F, T>(mut f: F) -> T
+fn with_thread_data<F, T>(f: F) -> T
 where
-    F: FnMut(&ThreadData) -> T,
+    F: FnOnce(&ThreadData) -> T,
 {
     // Try to read from thread-local storage, but return None if the TLS has
     // already been destroyed.
     #[cfg(has_localkey_try_with)]
-    fn try_with_tls<F, T>(key: &'static LocalKey<ThreadData>, f: &mut F) -> Option<T>
-    where
-        F: FnMut(&ThreadData) -> T,
-    {
-        key.try_with(|x| f(x)).ok()
+    fn try_get_tls(key: &'static LocalKey<ThreadData>) -> Option<*const ThreadData> {
+        key.try_with(|x| x as *const ThreadData).ok()
     }
     #[cfg(not(has_localkey_try_with))]
-    fn try_with_tls<F, T>(key: &'static LocalKey<ThreadData>, f: &mut F) -> Option<T>
-    where
-        F: FnMut(&ThreadData) -> T,
-    {
-        panic::catch_unwind(AssertUnwindSafe(|| key.with(|x| f(x)))).ok()
+    fn try_get_tls(key: &'static LocalKey<ThreadData>) -> Option<*const ThreadData> {
+        panic::catch_unwind(|| key.with(|x| x as *const ThreadData)).ok()
     }
 
+    #[allow(unused_assignments)]
+    let mut thread_data_storage = None;
+    let mut thread_data_ptr = ptr::null();
     // If ThreadData is expensive to construct, then we want to use a cached
     // version in thread-local storage if possible.
     if !cfg!(windows) && !cfg!(all(feature = "nightly", target_os = "linux")) {
         thread_local!(static THREAD_DATA: ThreadData = ThreadData::new());
-        if let Some(result) = try_with_tls(&THREAD_DATA, &mut f) {
-            return result;
+        if let Some(tls_thread_data) = try_get_tls(&THREAD_DATA) {
+            thread_data_ptr = tls_thread_data;
         }
     }
+    if thread_data_ptr.is_null() {
+        // Otherwise just create a ThreadData on the stack
+        thread_data_storage = Some(ThreadData::new());
+        thread_data_ptr = thread_data_storage.as_ref().unwrap();
```

Comment:
I don't see how that would work here. `thread_data_storage` is always `None` when reaching this if block. So the closure to `get_or_insert_with` would always run and I would still need to get the address out for my pointer variable later. My initial declaration of `thread_data_storage` was a bit off, but I moved it now to better show how it's used.

Reply:
```
thread_data_ptr = thread_data_storage.get_or_insert_with(ThreadData::new);
```
Since `get_or_insert_with` returns a reference to the inner value in the option, you can avoid calling `as_ref` and `unwrap`.

Reply:
:man_facepalming: I completely missed it returns a reference. Of course it does.